Move Docs From Wiki

CursedRock17 opened the pull request:
As per the title, this pull request moves the documentation of `message_filters` over to `docs.ros.org` from the WIKI, fully updated to fit current ROS 2 stadards. Furthermore, this pr contians new, in-depth tutorials for users to get better with `message_filters` and understand the various filters and policies. That part of the pull request comes from both [this `ros2_documentation` issue](https://github.com/ros2/ros2_documentation/issues/3218) and [this `ros2_documentation` issue](https://github.com/ros2/ros2_documentation/issues/2272) in which the tutorials needed an update. So, I added those tutorials over here so the docs would be centralized and only people using the package would have to worry about it, but if we needed, we can just copy all the information over to a PR in that repo. Additionally, I cannot get the ros2 docs to work for Python, though it works in C++, it comes from this line: https://github.com/CursedRock17/message_filters/blob/66c6bf9230c95cecc87c974500d759772a4e16de/doc/index.rst?plain=1#L263

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+and may or may not output the message
+at some time in the future, depending on a policy defined for that filter.
+
+message_filters also defines a common interface for these filters, allowing you to chain them together.
```
```suggestion
``message_filters`` also defines a common interface for these filters, allowing you to chain them together.
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+Input:
+  No input connections
+Output:
+  * C++: void callback(const std::shared_ptr<M const>&)
+  * Python: callback(msg)
```
```suggestion
  * C++: ``void callback(const std::shared_ptr<M const>&)``
  * Python: ``callback(msg)``
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+
+Output:
+  * C++: For message types M0..M8, void callback(const std::shared_ptr<M0 const>&, ..., const std::shared_ptr<M8 const>&). The number of parameters is determined by the number of template arguments the class was created with.
+  * Python: callback(msg0.. msgN). The number of parameters is determined by the number of template arguments the class was created with.
```
```suggestion
  * Python: ``callback(msg0.. msgN)``. The number of parameters is determined by the number of template arguments the class was created with.
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+4.1 Connections
+~~~~~~~~~~~~~~~
+Input:
+  * C++: void callback(const std::shared_ptr<M const>&)
```
```suggestion
  * C++: ``void callback(const std::shared_ptr<M const>&)``
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+Input:
+  * C++: void callback(const std::shared_ptr<M const>&)
+Output:
+  * C++: void callback(const std::shared_ptr<M const>&)
```
```suggestion
  * C++: ``void callback(const std::shared_ptr<M const>&)``
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+5.1 Connections
+~~~~~~~~~~~~~~~
+Input:
+  * C++: void callback(const std::shared_ptr<M const>&)
+  * Python: callback(msg)
+Output:
+  * C++: void callback(const std::shared_ptr<M const>&)
+  * Python: callback(msg)
```
```suggestion
  * C++: ``void callback(const std::shared_ptr<M const>&)``
  * Python: ``callback(msg)``
Output:
  * C++: ``void callback(const std::shared_ptr<M const>&)``
  * Python: ``callback(msg)``
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+    sub = message_filters.Subscriber('my_topic', sensor_msgs.msg.Image)
+    cache = message_filters.Cache(sub, 100)
+
+In this example, the Cache stores the last 100 messages received on my_topic, and myCallback is called on the addition of every new message. The user can then make calls like cache.getInterval(start, end) to extract part of the cache.
+If the message type does not contain a header field that is normally used to determine its timestamp, and the Cache is contructed with allow_headerless=True, the current ROS 2 time is used as the timestamp of the message. This is currently only available in Python.
```
```suggestion
In this example, the Cache stores the last 100 messages received on ``my_topic``, and ``myCallback`` is called on the addition of every new message. The user can then make calls like ``cache.getInterval(start, end)`` to extract part of the cache.
If the message type does not contain a header field that is normally used to determine its timestamp, and the Cache is contructed with ``allow_headerless=True``, the current ROS 2 time is used as the timestamp of the message. This is currently only available in Python.
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+----------------------------
+The Synchronizer filter synchronizes incoming channels by the timestamps contained in their headers, and outputs them in the form of a single callback that takes the same number of channels. The C++ implementation can synchronize up to 9 channels.
+
+The Synchronizer filter is templated on a policy that determines how to synchronize the channels. There are currently two policies: ExactTime and ApproximateTime.
```
There is a third one [ApproximateEpsilonTime](https://github.com/ros2/message_filters/blob/rolling/include/message_filters/sync_policies/approximate_epsilon_time.h). Not need to write a independent tutorial in C++ and Python, you can open an issue with this task

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+6.1 Connections
+~~~~~~~~~~~~~~~
+Input:
+  * C++: Up to 9 separate filters, each of which is of the form void callback(const std::shared_ptr<M const>&). The number of filters supported is determined by the number of template arguments the class was created with.
+  * Python: N separate filters, each of which has signature callback(msg).
+Output:
+  * C++: For message types M0..M8, void callback(const std::shared_ptr<M0 const>&, ..., const std::shared_ptr<M8 const>&). The number of parameters is determined by the number of template arguments the class was created with.
+  * Python: callback(msg0.. msgN). The number of parameters is determined by the number of template arguments the class was created with.
```
```suggestion
  * C++: Up to 9 separate filters, each of which is of the form ``void callback(const std::shared_ptr<M const>&)``. The number of filters supported is determined by the number of template arguments the class was created with.
  * Python: N separate filters, each of which has signature ``callback(msg)``.
Output:
  * C++: For message types M0..M8, ``void callback(const std::shared_ptr<M0 const>&, ..., const std::shared_ptr<M8 const>&)``. The number of parameters is determined by the number of template arguments the class was created with.
  * Python: ``callback(msg0.. msgN)``. The number of parameters is determined by the number of template arguments the class was created with.
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+
+6.2 ExactTime Policy
+~~~~~~~~~~~~~~~~~~~~
+The message_filters::sync_policies::ExactTime policy requires messages to have exactly the same timestamp in order to match. Your callback is only called if a message has been received on all specified channels with the same exact timestamp. The timestamp is read from the header field of all messages (which is required for this policy).
```
```suggestion
The ``message_filters::sync_policies::ExactTime`` policy requires messages to have exactly the same timestamp in order to match. Your callback is only called if a message has been received on all specified channels with the same exact timestamp. The timestamp is read from the header field of all messages (which is required for this policy).
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+
+6.3 ApproximateTime Policy
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+The message_filters::sync_policies::ApproximateTime policy uses an adaptive algorithm to match messages based on their timestamp.
+
+If not all messages have a header field from which the timestamp could be determined, see below for a workaround. If some messages are of a type that doesn't contain the header field, ApproximateTimeSynchronizer refuses by default adding such messages. However, its Python version can be constructed with allow_headerless=True, which uses current ROS 2 time in place of any missing header.stamp fiel ...
```
```suggestion
The ``message_filters::sync_policies::ApproximateTime`` policy uses an adaptive algorithm to match messages based on their timestamp.

If not all messages have a header field from which the timestamp could be determined, see below for a workaround. If some messages are of a type that doesn't contain the header field, ``ApproximateTimeSynchronizer`` refuses by default adding such messages. However, its Python version can be constructed with ``allow_headerless=True``, which uses current ROS 2 time in place of any missing header.stamp field:
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+
+7.1 Connections
+~~~~~~~~~~~~~~~
+Input:
+  * C++: void callback(const std::shared_ptr<M const>&)
+Output:
+  * C++: void callback(const std::shared_ptr<M const>&)
```
```suggestion
Input:
  * C++: ``void callback(const std::shared_ptr<M const>&)``
Output:
  * C++: ``void callback(const std::shared_ptr<M const>&)``
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+message filter (e.g. :class:`message_filters.Cache`) register their own
+message handler as a callback.
+
+Output connections are registered through the ``registerCallback()`` function.
+
+
+Here's a simple example of using a Subscriber with a Cache::
```
```suggestion
Output connections are registered through the ``registerCallback()`` function.

Here's a simple example of using a Subscriber with a Cache::
```

ahcorde commented on doc/index.rst:
```diff
@@ -0,0 +1,268 @@
+    cache = message_filters.Cache(sub, 10)
+    cache.registerCallback(myCallback)
+
+The Subscriber here acts as the source of messages.  Each message is passed to the cache, which then passes it through to the
+user's callback ``myCallback``.
```
```suggestion
The Subscriber here acts as the source of messages. Each message is passed to the cache, which then passes it through to the
user's callback ``myCallback``.
```

ahcorde commented on doc/Tutorials/Approximate-Synchronizer-Python.rst:
```diff
@@ -0,0 +1,155 @@
+
+For this example we will be using the ``temperature`` and ``fluid_pressure`` messages found in
+`sensor_msgs <https://github.com/ros2/common_interfaces/tree/rolling/sensor_msgs/msg>`_.
+To simulate a working ``ApproximateTimeSynchronizer``. We will be publishing and subscribing to topics of those respective types, to showcase how real sensors would be working. To simulate them we will also need two ``Timers`` on different intervals. Then, we will be utilizing an ``ApproximateTimeSynchronizer`` to get these messages from the sensor topics aligned with a slight delay between messa ...
+
+
+It is essential that the QoS is the same for all of the publishers and subscribers, otherwise the Message Filter cannot align the topics together. So, create one ``QoSProfile`` and stick with it, or find out what ``qos`` is being used in the native sensor code, and replicate it. Do basic construction of each object relating to the ``Node`` and callback methods that may be used in the future. Both ...
```
```suggestion
To simulate a working ``ApproximateTimeSynchronizer``. We will be publishing and subscribing to topics of those respective types, to showcase how real sensors would be working. To simulate them we will also need two ``Timers`` on different intervals. Then, we will be utilizing an ``ApproximateTimeSynchronizer`` to get these messages from the sensor topics aligned with a slight delay between messages..

It is essential that the QoS is the same for all of the publishers and subscribers, otherwise the Message Filter cannot align the topics together. So, create one ``QoSProfile`` and stick with it, or find out what ``qos`` is being used in the native sensor code, and replicate it. Do basic construction of each object relating to the ``Node`` and callback methods that may be used in the future. Both of the two timers we utilize will have different timer values of ``1`` and ``1.05`` which causes the timers to off at different points, which is an advantage of using ``ApproximateTime``. Notice that we must call ``sync->registerCallback`` to sync up the two (or more) chosen topics.
```

ahcorde commented on doc/Tutorials/Approximate-Synchronizer-Python.rst:
```diff
@@ -0,0 +1,155 @@
+
+3. Build
+~~~~~~~~
+From the root of your package, build and source.
+
+
+.. code-block:: bash
```
```suggestion
From the root of your package, build and source.

.. code-block:: bash
```

ahcorde commented on doc/Tutorials/Writing-A-Time-Synchronizer-Cpp.rst:
```diff
@@ -0,0 +1,191 @@
+
+3. Build
+~~~~~~~~
+From the root of your package, build and source.
+
+
+.. code-block:: bash
```
```suggestion
From the root of your package, build and source.

.. code-block:: bash
```

CursedRock17 commented:
I've added those necessary changes but when it comes to [this comment](https://github.com/ros2/message_filters/pull/119#discussion_r1604565909), we're looking to do a whole other Pull request for those tutorials after this one gets merged correct? Also, should I add a section for the `latest_time` policy as well?

ahcorde reviewed (approved):
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=21110)](http://ci.ros2.org/job/ci_linux/21110/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=15409)](http://ci.ros2.org/job/ci_linux-aarch64/15409/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=21824)](http://ci.ros2.org/job/ci_windows/21824/)

ahcorde reviewed (changes requested):
I can see these errors:

```bash
/home/ahcorde/ros2_rolling/src/ros2/message_filters/doc/Tutorials/Approximate-Synchronizer-Cpp.rst:1: ERROR: Unknown directive type "redirect-from".

.. redirect-from::

   Tutorials/Approximate-Synchronizer-Cpp
/home/ahcorde/ros2_rolling/src/ros2/message_filters/doc/Tutorials/Approximate-Synchronizer-Python.rst:1: ERROR: Unknown directive type "redirect-from".

.. redirect-from::

   Tutorials/Approximate-Synchronizer-Python
/home/ahcorde/ros2_rolling/src/ros2/message_filters/doc/Tutorials/Writing-A-Time-Synchronizer-Cpp.rst:1: ERROR: Unknown directive type "redirect-from".

.. redirect-from::

   Tutorials/Writing-A-Time_Synchronizer-Cpp
/home/ahcorde/ros2_rolling/src/ros2/message_filters/doc/Tutorials/Writing-A-Time-Synchronizer-Python.rst:1: ERROR: Unknown directive type "redirect-from".

.. redirect-from::

   Tutorials/Writing-A-Time_Synchronizer-Python
```

In the `index.html` I can't see any link to the tutorials or API links

CursedRock17 commented:
Yeah I messed up the `index.rst` file when I was dealing with the merge conflict, added in an extra one somehow. Turns out those redirects were useless so I removed those and created a singular, correct index file. Though now I've got all the updated code in the pull request, so I assume I need to remove in order to have a nice PR.

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=21168)](http://ci.ros2.org/job/ci_linux/21168/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=15473)](http://ci.ros2.org/job/ci_linux-aarch64/15473/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=21892)](http://ci.ros2.org/job/ci_windows/21892/)

CursedRock17 commented:
I don't know what I had done before with my rebase but it completely messed up the last few commits, all the linters pass now and we shouldn't have any CI problems, additionally all the tutorial links work

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=21170)](http://ci.ros2.org/job/ci_linux/21170/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=15475)](http://ci.ros2.org/job/ci_linux-aarch64/15475/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=21894)](http://ci.ros2.org/job/ci_windows/21894/)
